### PR TITLE
Piercetrey figure/sc 253074/update to blockapi

### PR DESCRIPTION
**Pull request by piercetrey-figure**
* Using blockapi for the event listening instead of the event stream
* Updated to Java 17

### CODE REVIEW

File: server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    private suspend fun blockStreamLoop() {
+        val lastBlockProcessed = blockHeightRepository.getLastProcessedBlockHeight()
+        blockAPIClient.streamBlocks(
+            start = lastBlockProcessed,
+            preference = BlockServiceOuterClass.PREFER.TX_EVENTS
+        ).collect { block ->
+            val lastProcessedHeight = blockHeightRepository.getLastProcessedBlockHeight()
+
+            block.blockResult.block.transactionsList.flatMap { it.eventsList }
```
Comment (piercetrey-figure) [inline server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt:61]: took me a while to figure out that block.blockResult.transactions.transactions.transactionsList was an empty red herring 🤦

File: .github/workflows/build-and-publish.yml [hunk 2]
```diff
@@ -28,10 +28,10 @@ jobs:
 
           echo "VERSION=$VERSION" >> $GITHUB_ENV
 
-      - name: Set up JDK 11
+      - name: Set up JDK 17
         uses: actions/setup-java@v1
         with:
-          java-version: 11
+          java-version: 17
           server-id: github # Value of the distributionManagement/repository/id field of the pom.xml
```
Comment (mtps) [inline .github/workflows/build-and-publish.yml:35]: I think you need to specify azul?

File: server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt (hunk 1 above)
Comment (mtps) [reply server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt:61]: Yea, there's a branch out there to nix that out that's about 8 months old.

File: .github/workflows/build-and-publish.yml (hunk 2 above)
Comment (piercetrey-figure) [reply .github/workflows/build-and-publish.yml:35]: looks like zulu

Comment (piercetrey-figure) [reply .github/workflows/build-and-publish.yml:35]: specified

**Review by mtps — APPROVED** [review]

**Comment by piercetrey-figure** [comment]
going to adjust this to be pluggable between event stream and blockapi since this is public and might need to be more flexible

File: README.md [hunk 3]
```diff
@@ -143,7 +143,13 @@ the records associated with a Provenance Blockchain Scope.
...
 
 #### Wasm Event Revokes
-This service watches the event stream emitted by the Provenance Blockchain using the [Figure Tech Event Stream Library](https://github.com/FigureTechnologies/event-stream).
+This service watches the event stream emitted by the Provenance Blockchain using either the
+[Figure Tech Event Stream Library](https://github.com/FigureTechnologies/event-stream)
+or the [Figure Tech Blockapi Library](https://github.com/FigureTechnologies/pb-block-client).
+The choice of stream can be configured using the blockstream.type property (or env var BLOCKSTREAM_TYPE),
+where the value "provenance" allows connecting to a Provenance node with the [Figure Tech Event Stream Library](https://github.com/FigureTechnologies/event-stream),
+and a value of "blockapi" allows connecting to the more efficient [Figure Tech Blockapi Library](https://github.com/FigureTechnologies/pb-block-client). Blockapi usage requires
+an api key for consumption.
```
Comment (hyperschwartz) [inline README.md:152]: Might be good to note the env var required for blockapi credentials here

File: gradle/libs.versions.toml [hunk 4]
```diff
@@ -11,13 +11,17 @@ hikari = "5.0.1"
 jackson = "2.12.5"
 jackson-protobuf = "0.9.13"
 jwt = "3.19.4"
-kotlin = "1.6.10"
-kotlin-coroutines = "1.5.2"
+kotlin = "1.9.0"
+kotlin-coroutines = "1.7.3"
+# IMPORTANT: Kotlin Logging cannot be upgraded to version 3.x+ because Spring relies on an older set of logging libraries
+# and the newer Kotlin Logging libraries mandate newer logging.  Spring will start exploding when the app launches in
+# kube due to the slf4j remover.
```
Comment (hyperschwartz) [inline gradle/libs.versions.toml:18]: :scumbag-springboot:

File: server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    @Qualifier(BeanQualifiers.BLOCK_STREAM_COROUTINE_SCOPE_QUALIFIER) private val blockStreamScope: CoroutineScope,
+) : ApplicationListener<ApplicationReadyEvent>, HealthIndicator {
+    private companion object : KLogging() {
+        private var blockStreamRunning = AtomicBoolean(false)
+        private val TX_EVENTS = setOf("wasm")
+    }
+
+    override fun onApplicationEvent(event: ApplicationReadyEvent) {
+        if (!blockStreamProperties.enabled) {
+            logger.warn("Event stream has been manually disabled! Use a value of `true` for EVENT_STREAM_ENABLED to re-enable it")
```
Comment (hyperschwartz) [inline server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt:38]: This probably needs to be changed to talk about blockapi

File: server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    override fun health(): Health = if (blockStreamRunning.get()) {
+        Health.up().build()
+    } else {
+        Health.down().build()
+    }
+
+    private suspend fun blockStreamLoop() {
+        val lastBlockProcessed = blockHeightRepository.getLastProcessedBlockHeight()
```
Comment (hyperschwartz) [inline server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt:53]: This is really cool how you can just mod your service on the fly with a redeploy to switch between the ES and Blockapi.  <3

File: server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            val lastProcessedHeight = blockHeightRepository.getLastProcessedBlockHeight()
+
+            block.events
+                .filter { TX_EVENTS.contains(it.eventType) } // these are the blocks you are looking for
+                .forEach {
+                    try {
+                        streamEventHandlerService.handleEvent(it)
+                    } catch (e: Exception) {
+                        // If exceptions are simply thrown without any additional logging, the errors appear to be
+                        // event stream related, which would not be the case if they occur in this block
```
Comment (hyperschwartz) [inline server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt:66]: This comment might not be true when regarding this service

File: server/src/main/kotlin/tech/figure/objectstore/gateway/configuration/BlockStreamConfig.kt [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import tech.figure.block.api.proto.BlockServiceOuterClass
+import tech.figure.objectstore.gateway.eventstream.BlockApiGatewayEventStream
+import tech.figure.objectstore.gateway.eventstream.GatewayEventStream
+import tech.figure.objectstore.gateway.eventstream.ProvenanceGatewayEventStream
+
+@Configuration
+class BlockStreamConfig {
+
+    @Bean
+    fun gatewayEventStream(blockStreamProperties: BlockStreamProperties): GatewayEventStream = when (blockStreamProperties.streamType) {
```
Comment (hyperschwartz) [inline server/src/main/kotlin/tech/figure/objectstore/gateway/configuration/BlockStreamConfig.kt:18]: This is a really cool way to do this

File: server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt (hunk 5 above)
Comment (hyperschwartz) [reply server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt:38]: JK i read further

File: server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt (hunk 7 above)
Comment (hyperschwartz) [reply server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt:66]: JK also

**Review by hyperschwartz — APPROVED** [review]

File: gradle/libs.versions.toml (hunk 4 above)
Comment (piercetrey-figure) [reply gradle/libs.versions.toml:18]: I just noticed group manager had this comment from you so it seemed important to bring over 😆

**Review by mtps — DISMISSED** [review]

File: server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt (hunk 5 above)
Comment (piercetrey-figure) [reply server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt:38]: no, that one is a good comment, need to update the name of the env var (also just noticed the readme listing of env vars that I need to update)

**Review by hyperschwartz — APPROVED** [review]

**Pull request merged**